homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -345,6 +346,7 @@
     ATTR_DRY: "dry",
     ATTR_FAHRENHEIT: "fahrenheit",
     ATTR_MOTOR_SPEED: "motor_speed",
+    ATTR_WATER_LEVEL: "water_level",
```

Comment:
Water level should be a separate sensor entity.

See https://github.com/home-assistant/core/pull/39398#discussion_r492523276

Please update the docs if needed instead.

Reply:
While I generally expected all of these to be separate sensors, I don't think I fully understand where one draws the line. For example, in case of this integration, I would have expected `water_level` to be a separate sensor, but the same would apply to `mode`, `child_lock`, `motor_speed`, `target_speed`, `humidity` and so on. Other devices (i.e. air purifier) also seem to implement this as state attributes rather than separate entities.

Is there some rule of thumb that makes `water_level` special here?

Reply:
@skrobul if the attribute is non-static meaning it can update outside the state of the entity then yes it should be a separate entity.  What you see for the other devices is older code that was accepted at one point in time but now anything new added needs to follow this rule of thumb.

Reply:
Yes. If it's just a measurement it should be a separate sensor entity. If it's a setting that can be controlled with an entity service, then we expose it as an attribute on the entity that supports the service.

Reply:
Thank you both. I have started reading and rereading Developers documentation and it seems like what you ask will be relatively large change and I don't know how to do all of it yet. 

Since the humidifier implementation uses only `fan` platform right now, I am planning to add the code into `sensor` platform which currently does not have any humidifier related code. 
In fact, I was able to write that part but it's very ugly now. It establishes brand new connection for every single entity (I implemented only `sensor.humidity` and `sensor.temperature` for testing and it works but in my mind, it is unacceptable to have a separate connection for every measurement. [DataUpdateCoordinator](https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities) sounds like a reasonable solution to try.

I have two questions though:
1. Is it OK to share data between platforms (presumably yes, but just checking) ? I.e. in this specific case, one call to lower level library (miio) will provide data for both `sensor`,  `fan` and `switch`. 
2. What is the deprecation policy if I was to implement entities for most of the stuff that is currently presented to the users as state attributes? Do they have to remain available as duplicates of newly created entities?

Reply:
The platforms should share the same connection to the device but a platform should not share platform code with another platform. So the connection needs to be set up from the init module when setting up the integration. Then the platforms can get access to this connection, eg by using the data update coordinator, if it's a polling implementation.

Reply:
In this case I think we can make a breaking change as long as the same measurements are made available in another way.